### PR TITLE
Wait for all background threads to finish before filling installation…

### PR DESCRIPTION
**Pull request by M4rtinK**
… task queue

Otherwise installation tasks, such as the installation task that
transfers subscription tokens, will be scheduled based on stale data
(eq. system is not subscribed as the subscription thread is still
running).

By waiting for all background processing threads to finish before
assembling the installation task queue, we will make sure all tasks are
correctly scheduled based on up to data data - eq. system was
successfully registered, so the token transfer task needs to transfer
the relevant subscription tokens to the target system.

Resolves: rhbz#1824422

### CODE REVIEW

File: pyanaconda/installation.py [hunk 1]
```diff
@@ -234,22 +234,15 @@ def _prepare_installation(payload, ksdata):
...
-            for message in ("Thread %s is running" % n for n in threadMgr.names):
-                log.debug(message)
+        for message in ("Thread %s is running" % n for n in threadMgr.names):
+            log.debug(message)
             threadMgr.wait_all()
-
-        # Use a queue with a single task as only TaskQueues have the status_message
-        # property used for setting the progress status in the UI.
-        wait_for_threads = TaskQueue("Wait for threads to finish",
-                                     N_("Waiting for %s threads to finish") % (threadMgr.running - 1))
```
Comment (poncovka) [inline pyanaconda/installation.py:249]: I think we should call `progressQ.send_message` before the waiting to show the same message on the progress screen as before.

Comment (M4rtinK) [reply pyanaconda/installation.py:249]: Good point, added - as far as I can tell, the output now looks similar to what was shown before, possibly just in different order due to the insufficient synchronization before.

**Review by poncovka — APPROVED** [review]
Looks good to me. Thanks!

File: pyanaconda/installation.py [hunk 2]
```diff
@@ -234,22 +235,17 @@ def _prepare_installation(payload, ksdata):
...
+    # still running at task creation time.
     if threadMgr.running > 1:
-        # it could be that the threads finish execution before the task is executed,
-        # but that should not cause any issues
-
-        def wait_for_all_treads():
-            for message in ("Thread %s is running" % n for n in threadMgr.names):
-                log.debug(message)
+        # show a progress message
+        progressQ.send_message(N_("Waiting for %s threads to finish") % (threadMgr.running - 1))
```
Comment (jkonecny12) [inline pyanaconda/installation.py:244]: I'm not sure if this is not a too technical message for a user. Maybe `Waiting for %s background jobs to finish` would be less technical? Not sure really.

File: pyanaconda/installation.py [hunk 3]
```diff
@@ -234,22 +235,17 @@ def _prepare_installation(payload, ksdata):
...
+        for message in ("Thread %s is running" % n for n in threadMgr.names):
```
Comment (jkonecny12) [inline pyanaconda/installation.py:245]: I would also change the message here to something like `Waiting on thread %s to finish` to know why is the information important.

**Review by jkonecny12 — APPROVED** [review]
Looks good to me. Just a bit improvement on messaging. However, I won't insist on that.

File: pyanaconda/installation.py [hunk 4]
```diff
@@ -234,22 +235,17 @@ def _prepare_installation(payload, ksdata):
...
+            log.debug(message)
             threadMgr.wait_all()
```
Comment (rvykydal) [inline pyanaconda/installation.py:247]: It seems to me that waiting for the methods should go out of the for loop ?

Comment (rvykydal) [inline pyanaconda/installation.py:247]: It seems to me that waiting for the threads should go out of the for loop  ?

Comment (poncovka) [reply pyanaconda/installation.py:247]: It should. Good catch!

File: pyanaconda/installation.py (hunk 2 above)
Comment (poncovka) [reply pyanaconda/installation.py:244]: These are the same messages as before. Do we really want to change them in RHEL 8?

Comment (M4rtinK) [reply pyanaconda/installation.py:244]: Agreed with @poncovka, I think we want to keep this externally consistent, at least on the RHEL branches.

Comment (jkonecny12) [reply pyanaconda/installation.py:244]: OK, good point.

File: pyanaconda/installation.py (hunk 4 above)
Comment (M4rtinK) [reply pyanaconda/installation.py:247]: Oh, right - I somehow thought this is intentional (and this code was the same before) and `wait_all()` returns every time a thread terminates, so you get a nice listing of threads as they terminate. But nope, `wait_all()` does indeed wait for all currently running non-main threads, so it really is pointless to all it like this. :D Lets fix that. :)

Comment (M4rtinK) [reply pyanaconda/installation.py:247]: Looking even more closely, the old method did not have `wait_all()` in the for cycle - double oops. :P

Comment (M4rtinK) [reply pyanaconda/installation.py]: Anyway should be fixed now. ;-)

**Review by rvykydal — APPROVED** [review]

**Review by VladimirSlavik — APPROVED** [review]
Looks good to me. Thank you!

**Review by poncovka — APPROVED** [review]

File: pyanaconda/installation.py [hunk 5]
```diff
@@ -234,22 +235,17 @@ def _prepare_installation(payload, ksdata):
     installation_queue.task_completed.connect(lambda x: progress_step(x.name))
 
     # This should be the only thread running, wait for the others to finish if not.
```
Comment (poncovka) [inline pyanaconda/installation.py:237]: In upstream, could you move this code to a new function and call it called from `run_installation`, please?

Comment (M4rtinK) [reply pyanaconda/installation.py:237]: Sounds good, let's do that in the upstream version. :)

Comment (M4rtinK) [reply pyanaconda/installation.py:237]: changed in upstream PR: #3720

**Comment by M4rtinK** [comment]
Port to Fedora: #3720

**Comment by M4rtinK** [comment]
Port to RHEL 9: #3722

**Pull request merged**